### PR TITLE
Version 2.0.0

### LINKED ISSUE #2: support export/import of data structure
due to the centroid implementation of the `TDigest` interface being private, it is currently not possible to get at the fields easily.

the use case I have in mind is creating and populating `centroidSet` on various separate machines, serializing them, sending them across the network to an aggregator machine, and then on this aggregator, deserializing and merging them.

### PR DESCRIPTION
**Pull request by spenczar**
#2 raised the point that, while TDigests can be merged, they can't be serialized, so merging isn't very practically useful.

Adding binary marshaling and unmarshaling to the *centroidSet struct makes the API unwieldy unless it's added to the TDigest interface too, but doing so is an API-breaking change.

I definitely want TDigest serialization and I regret the interface-based API, so this is a reasonable opportunity to do a major version bump.

So, this revision adds binary serialization, and retypes TDigest as a struct.

Fixes #2.

### CODE REVIEW

**Comment by spenczar** [comment]
@igorwwwwwwwwwwwwwwwwwwww, could you take a look and see whether this meets your needs?

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
@spenczar wow, thank you for the fast response! this is _exactly_ what I'm looking for. :)

File: tdigest.go [hunk 1]
```diff
@@ -22,10 +25,17 @@ import (
 //
 // MergeInto(other) will add all of the data within a TDigest into
 // other, combining them into one larger TDigest.
+//
+// Finally, TDigests implement the encoding package's BinaryMarshaler and
```
Comment (rhysh) [inline tdigest.go]: Changing the interface by adding these methods is a breaking change—if other implementations don't happen to have these new methods, their builds will fail.

File: tdigest.go [hunk 2]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
...
+	)
+	r := &stickyErrReader{r: bytes.NewReader(p), err: nil}
+
+	binary.Read(r, binary.LittleEndian, &encodingVersion)
+	if encodingVersion != 1 {
+		return fmt.Errorf("invalid encoding version: %d", encodingVersion)
+	}
+
+	binary.Read(r, binary.LittleEndian, &min)
+	binary.Read(r, binary.LittleEndian, &max)
```
Comment (rhysh) [inline tdigest.go]: The values of `min` and `max` aren't used here. Do they need to be part of the serialization?

File: tdigest.go [hunk 3]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
 		other.Add(c.mean, c.count)
 	}
 }
+
+func (cs *centroidSet) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	w := &stickyErrWriter{w: buf, err: nil}
+	binary.Write(w, binary.LittleEndian, int32(1))
```
Comment (rhysh) [inline tdigest.go]: Please label this magic number. It could be a line within this function of `const encodingVersion int32 = 1`.

File: tdigest.go [hunk 4]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
 		other.Add(c.mean, c.count)
 	}
 }
+
+func (cs *centroidSet) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	w := &stickyErrWriter{w: buf, err: nil}
```
Comment (rhysh) [inline tdigest.go]: I really like that you made a type to help here. The boilerplate of `binary.Write(w, binary.LittleEndian, v)` remains; can the custom type encompass that as well, exposing a method like `WriteValue(v interface{})`?

File: tdigest.go [hunk 5]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
...
+
+	binary.Write(w, binary.LittleEndian, min)
+	binary.Write(w, binary.LittleEndian, max)
+	binary.Write(w, binary.LittleEndian, float64(cs.compression))
+	binary.Write(w, binary.LittleEndian, int32(len(cs.centroids)))
+	if w.err != nil {
+		return nil, w.err
+	}
+	for _, c := range cs.centroids {
+		binary.Write(w, binary.LittleEndian, int64(c.count))
```
Comment (rhysh) [inline tdigest.go]: I really like that this is an `int64`. Consider changing the types of `centroidSet.countTotal` and `centroid.Count` to `int64` rather than `int`—though maybe that should happen as a separate maintenance commit.

File: tdigest.go [hunk 6]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
...
+		binary.Write(w, binary.LittleEndian, float64(c.mean))
+	}
+	if w.err != nil {
+		return nil, w.err
+	}
+	return buf.Bytes(), nil
+}
```
Comment (rhysh) [inline tdigest.go]: The `centroidSet.reclusterAt` field doesn't get marshaled here, and it appears to be unused in the whole package. Consider removing it, possibly in a separate maintenance commit.

File: tdigest.go [hunk 7]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
...
+	if len(cs.centroids) > 0 {
+		min = cs.centroids[0].mean
+		max = cs.centroids[len(cs.centroids)-1].mean
+	}
+
+	binary.Write(w, binary.LittleEndian, min)
+	binary.Write(w, binary.LittleEndian, max)
+	binary.Write(w, binary.LittleEndian, float64(cs.compression))
+	binary.Write(w, binary.LittleEndian, int32(len(cs.centroids)))
+	if w.err != nil {
```
Comment (rhysh) [inline tdigest.go]: Under what circumstances would this have an error? I think we know statically that there can't be any, so why not check only a single time (at the end)?

File: tdigest.go [hunk 8]
```diff
@@ -335,3 +345,96 @@ func (cs *centroidSet) MergeInto(other TDigest) {
...
+		var (
+			count int64
+			mean  float64
+		)
+		binary.Read(r, binary.LittleEndian, &count)
+		binary.Read(r, binary.LittleEndian, &mean)
+		if r.err != nil {
+			return r.err
+		}
+		cs.centroids[i] = &centroid{count: int(count), mean: mean}
```
Comment (rhysh) [inline tdigest.go]: Does this package have invariants that can be broken by unmarshaling bad/crafted input? If those can lead to a panic later (I haven't examined that), we should check the invariants here and return an error.

**Review by rhysh — CHANGES_REQUESTED** [review]
I have some cleanup requests, but the main thing to discuss is the API-breaking change. It might be possible to get away without breaking the API, but if a change is inevitable then let's consider other possible API changes (like returning a pointer to an exported struct).

Have you considered fuzzing this new code to search for crashes, if bad input breaks an invariant?

**Comment by spenczar** [comment]
Great point that this is API breaking anyway. It would be nice to move to an exported struct and drop the interface - providing an interface is pretty clumsy for upgrades, and it adds nothing to this package. I'll take a look at that (and all the other comments, thanks for the review!)

File: tdigest.go (hunk 4 above)
Comment (spenczar) [reply tdigest.go]: Done, good idea.

File: tdigest.go (hunk 3 above)
Comment (spenczar) [reply tdigest.go]: Yep, done

File: tdigest.go (hunk 7 above)
Comment (spenczar) [reply tdigest.go]: Good point, removed this check.

File: tdigest.go (hunk 5 above)
Comment (spenczar) [reply tdigest.go]: Right, done in this revision. I expect this to be v2.0.0 when it lands anyway.

File: tdigest.go (hunk 6 above)
Comment (spenczar) [reply tdigest.go]: Thanks, removed it.

File: tdigest.go (hunk 2 above)
Comment (spenczar) [reply tdigest.go]: I'm just matching the behavior of the java reference implementation's serialization, here.

File: tdigest.go (hunk 8 above)
Comment (spenczar) [reply tdigest.go]: Maybe, but I don't think so. I'll try fuzzing.

File: tdigest.go (hunk 1 above)
Comment (spenczar) [reply tdigest.go]: Right, good point. I've taken this opportunity to clean up the API and return a struct. This will be v2.0.0.

**Comment by rhysh** [comment]
The changes look good at first glance, but it's hard to follow the new code vs the refactoring/API change. Can you split it up into smaller pieces? There might be four of them, but each would be really easy to read.

Maybe 1) remove TDigest interface, return a pointer to the TDigest struct, 2) change some `int` fields to `int64` and remove some unused fields, 3) rename every `cs` variable to `d`, 4) add marshal/unmarshal. I think those could happen in basically any order, since packages like encoding/json should be able to discover the methods even if the type isn't exported.

**Comment by spenczar** [comment]
@rhysh, yes, sorry about the mess. I'd like them all to be in this PR, though, so that I can squash them into just one API-breaking commit. I'll add a revert commit, and then add the four commits you described.

**Comment by spenczar** [comment]
@rhysh, please take another look. I've split things into commits that should be easier to read, and I was careful to make sure that tests passed at each commit.

**Comment by rhysh** [comment]
Thanks for splitting it up, each commit was much easier to read.

I'm not sold on the idea of compatibility with the Java implementation, particularly since it's not verified in a test. I'll defer to your judgement there. 1) Can you please include a permalink (i.e. including a commit id) to the Java counterpart? I think it got lost in the shuffle.

It's great to hear that go-fuzz found stuff! 2) Can you include some calls to Add and Quantile, and possibly MergeInto so the fuzzing also covers any invariants in the rest of the package? "What if a centroid.count is negative" comes to mind. 3) If it's easy, can you convert the fuzz regression cases to (printable) ASCII? "fuzz2" includes some printable non-ASCII.

4) Using int64 in the Add method could make it harder to use, and passing huge (1<<31 and up) values there seems unlikely. I'm on the fence—how happy are you with the signature?

**Comment by spenczar** [comment]
> I'm not sold on the idea of compatibility with the Java implementation, particularly since it's not verified in a test. I'll defer to your judgement there.

Yeah, that's a fair point. And indeed, we only support one of the two encodings - not that either is documented anywhere other than in code. I think you're right, overall - I'll trim things down.

> Can you include some calls to Add and Quantile, and possibly MergeInto so the fuzzing also covers any invariants in the rest of the package? "What if a centroid.count is negative" comes to mind.

Sure... I know quite a few invariants that could be violated.

> If it's easy, can you convert the fuzz regression cases to (printable) ASCII? "fuzz2" includes some printable non-ASCII.

Sure.

> Using int64 in the Add method could make it harder to use, and passing huge (1<<31 and up) values there seems unlikely. I'm on the fence—how happy are you with the signature?

Hm, I hadn't thought about this much. It seems kind of odd to store the value differently than we receive it... but I think you're right that it shouldn't matter. I'll change it.

File: serde.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+				return fmt.Errorf("data corruption detected: centroid %d has lower mean (%v) than preceding centroid %d (%v)", i, c.mean, i-1, prev.mean)
+			}
+		}
+		d.centroids[i] = c
+		if c.count > math.MaxInt64-d.countTotal {
+			return fmt.Errorf("data corruption detected: centroid total size overflow")
+		}
+		d.countTotal += c.count
+	}
+
```
Comment (rhysh) [inline serde.go:90]: Should there be an error if the provided `[]byte` contains extra trailing bytes? The `encoding.BinaryUnmarshaler` don't really say, but it seems good to be strict here.

**Review by rhysh — APPROVED** [review]
This change LGTM as-is, and would LGTM with a change to error on trailing bytes (if you think that's right).

File: serde.go (hunk 9 above)
Comment (spenczar) [reply serde.go:90]: Yeah, probably. Will do.
